frontend/lib/resolution/scope-queries.cpp

```diff
@@ -562,18 +634,33 @@ static bool doLookupInScope(Context* context,
     if (onlyInnermost && got) return true;
   }
 
+  // Consider the receiver scopes directly (for fields).
+  // Only consider receiver scopes if we are at a method scope
+  // in order to get field vs. formal precedence correct.
+  // (other cases will be handled later in this function)
+  if (considerReceiverScopes && scope->isMethodScope()) {
+    // create a config that doesn't search receiver scopes parent scopes
+    // (such parent scopes are covered in later in this function)
+    LookupConfig newConfig = config;
+    newConfig &= ~(LOOKUP_PARENTS|LOOKUP_GO_PAST_MODULES|LOOKUP_TOPLEVEL);
+    // and only consider methods/fields
+    // (but that's all that we should find in a record/class decl anyway...)
+    newConfig |= LOOKUP_ONLY_METHODS_FIELDS;
 
-  if (checkParents) {
-    LookupConfig newConfig = LOOKUP_DECLS;
-    if (checkUseImport) {
-      newConfig |= LOOKUP_IMPORT_AND_USE;
-    }
-    if (skipPrivateVisibilities) {
-      newConfig |= LOOKUP_SKIP_PRIVATE_VIS;
-    }
-    if (onlyInnermost) {
-      newConfig |= LOOKUP_INNERMOST;
+    bool got = false;
+    for (const auto& currentScope : receiverScopes) {
+      got |= doLookupInScope(context, currentScope, {}, resolving,
+                             name, newConfig, checkedScopes, result);
     }
+    if (onlyInnermost && got) return true;
```

Comment:
I wrote it this way on purpose. I was expecting that, if there are duplicate symbols with the same name among a class and its superclasses, we would want an ambiguity error in this case.

Looking at the production compiler, we get an error about `field has the same name as parent class field`:

``` chapel
module M {
  class Parent {
    var field: int;
  }
  class Child : Parent {
    var field: real;
  }

  proc Child.foo() {
    return field;
  }

  var c = new Child();
  var x = c.foo();
  writeln(x);
}
```

Ditto for a nested type:

``` chapel
module M {
  class Parent {
    record RR { var field: int; }
  }
  class Child : Parent {
    record RR { var field: real; }
  }

  proc Child.foo() type {
    return RR;
  }

  var c = new Child();
  type t = c.foo();
  var x: t;
  writeln(x);
}
```

Right now, the dyno compiler isn't emitting an error for these cases, but it also isn't succeeding at resolving the `foo` method. Vs., if we put the exit inside of the for loop, it will resolve these to the Child.field / Child.RR. I think we will need to improve the errors in these cases, but that's a separate effort from this PR & I think what I have in the PR here is closer to the behavior we want in the end. Note that the production compiler used to prefer the child class / subclass fields, but see #19505 where we changed that to instead disallow fields with the same name as a parent class / superclass field.

Reply:
Sounds good, thanks for the explanation.